src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs

```diff
@@ -357,13 +357,13 @@ private static void RemoveDirectoryRecursive(string fullPath, ref Interop.Kernel
                         else
                         {
                             // Name surrogate reparse point, don't recurse, simply remove the directory.
-                            // If a mount point, we have to delete the mount point first.
+                            // If a volume mount point, we have to delete the mount point first.
                             if (findData.dwReserved0 == Interop.Kernel32.IOReparseOptions.IO_REPARSE_TAG_MOUNT_POINT)
                             {
                                 // Mount point. Unmount using full path plus a trailing '\'.
                                 // (Note: This doesn't remove the underlying directory)
                                 string mountPoint = Path.Join(fullPath, fileName, PathInternal.DirectorySeparatorCharAsString);
-                                if (!Interop.Kernel32.DeleteVolumeMountPoint(mountPoint) && exception == null)
+                                if (PathInternal.IsDevice(GetImmediateLinkTarget(mountPoint, true, false, true)) && !Interop.Kernel32.DeleteVolumeMountPoint(mountPoint) && exception == null)
```

Comment:
This won't be filtering juntions from volume mount points. Both will return without device path because GetImmediateLinkTarget always removes the `\??\` prefix. 

Which means that we will stop using `DeleteVolumeMountPoint` and will always be using `RemoveDirectory` for actual volume mount points. Surprisingly, your test doesn't fail due to this but we should avoid intrudicing the risk.